commonknowledge/mapped
# Better address geocoding [CC-133]

Moves address geocoding from OSM to Google, for better fuzzy matching on place names.

## Description
- Moves from OSM's free API to Google's paid one
- Adds a `countries` column to data sources as a guardrail for very very fuzzy addresses. No UI for now, but can add this when we start onboarding international orgs.
- Adds geocode caching
- Adds postcode_data to geocoded addresses, so that all locations have a standardised set of political geography data
- Adds some tests for all our geo queries to postcodes IO
- (Adds an unused Mapbox implementation of this, in case we need to make a quick switch due to financial constraints)

## Motivation and Context

Most data processed through maps is postcode data. However for the Climate Coalition's hub, there are lists of events that have only fuzzy address names. And in future, when adding lists of researched targets, it's again likely that address data will come unstructured.

So this PR implements Google's geocoding API for `address`-based sources. I chose Google because it has the most accurate matching for place names, like "Pelican House", or fuzzy / local names for locations like art centres and churches. We would expect organisers to use locally prevalent names. Unfortunately Mapbox fared just a bit worse when guessing based on this.

Financially, although the Google API is paid, (£500 for 100,000 records), we also get £200 starter credits. Both Google and Mapbox charge £500 per 100,000 stored records. Just in case things go haywire, [I've added a policy to alert us if over 1,000 geocoding requests are made in a day.](https://console.cloud.google.com/monitoring/alerting/policies/6990678199335680843?project=mapped-v2&supportedpurview=project)

## How Can It Be Tested?

0. Get your API key from Google Cloud Console (I've invited all developers) or ask Jan for a key
1. Add a new data source with some addresses like "London" or "Pelican House" or "Glasgow Central Mosque" and the `ADDRESS` type
2. Import the data
3. You should see it all looks fine on the map

## How Will This Be Deployed?

- [x] Add Google Maps API to production env
